rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+introduce new build tools.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+Platform support:
+  * Linux - required, I presume
+  * MacOS - my development environment
+  * Windows - this is a policy decision.  I can design for
+    portability, but I do not have a Windows machine so I cannot test
+    Windows support.  I do have a copy of Parallels, so if somebody
+    can provide me with a licensed copy of Windows I can support it.
+
+### Tasks
+
+General:
+* Minimal build (local monobuild).
+  * WORKSPACE file, BUILD.bazel files containing target specifications, .bzl files containing macros, global vars, custom rules, etc.
+  * `$ bazel build //pkg:target` completes successfully for each target
+  * MacOS first, then Linux, then Windows.
+* Adaptive builds (local monobuilds with platform-specific adaptations).
+  * A Bazel `platform()` is defined for each officially supported platform
+    * Platform definition includes architecture, OS, toolchain.
+  * Build specifications are adapted to each supported platform
+    * For example, the compilers for different platforms take
+      different options.  Bazel does a good job of automatically
+      setting the compile command for each platform, but in some cases
+      additional options should or must be specified.
+  * Bazel does not have `./configure` functionality; it cannot detect
+    environment features, at least not out of the box.  Such
+    functionality could be implemented in Bazel's Starlark language,
+    but that is outside the scope of this proposal.  In the meantime,
+    Bazel works well with `./configure`; if we need feature detection
+    (e.g. to support all three major platforms), a configure script
+    can be written to do this, just as in the case of a Make-based
+    system.
+* Toolchains.
+  * Bazel detects and uses a standard set of commonly used toolchains; no effort required.
+  * For toolchains not covered by Bazel out of the box, as well as for
+    cross-build toolchains, a set of rules defining the toolchain for
+    Bazel must be developed.
+    * A .bzl file containing code describing the location of tools, API, features, etc.
+    * To make the defined toolchain known to Bazel, a BUILD file
+      containing rules associating the toolchain with build properties, thus enabling build-time toolchain resolution
+
+Specific tasks:
+* C/C++
+  * External libs. Most of these are already bazelized (minimal build on MacOS).
+    * jemalloc, libffi, libgmp, libomp, libpatch, libpq, libprocps, libre2, libsodium, openssl, zlib
+  * libsnark.
+    * A special case.  Coda uses a modified fork of libsnark; it
+      depends on a handful of C/C++ libs (libff, libfqfft, etc.).
+    * I have forked the dependencies on github and added Bazel support
+      to each (about 90% complete).  The ideal is to get the
+      maintainers to accept the Bazel-enabling patches.
+    * For Coda this approach may or may not be desirable.  The
```

Comment:
While we wait for patches to be upstreamed on the internal deps and external BUILD file seems like a good option.